archive_upload/handlers/dsmc_handlers.py

```diff
@@ -736,19 +736,43 @@ def exclude_content(tarinfo):
 
         # Remove files that we added to the tarball.
         with tarfile.open(tarball_path) as tar:
-            for member in tar.getmembers():
-                try:
-                    filepath = os.path.normpath(
-                        os.path.join(path_to_archive_root, archive, member.name))
-
-                    if os.path.isfile(filepath):
-                        os.remove(filepath)
-                    elif member.name != "." and os.path.isdir(filepath):
-                        self._rm_empty_dirs(filepath)
-
-                except OSError, e:
-                    msg = "Error when creating archive tarball. Could not remove file {}: {}".format(filepath, e)
-                    raise ArchiveException(reason=msg, status_code=500)
+            # restore the original paths for the members in the tarball
+            original_member_paths = map(
+                lambda m: os.path.normpath(os.path.join(path_to_archive, m.name)),
+                tar.getmembers())
+            # walk down the archive (do not follow links!) and remove links that have been included in the tarball
+            # this way seems safer than just iterating over the members in the tarball
+            files_to_remove = []
+            dirs_to_remove_if_empty = []
+            for dirpath, subdirs, dirfiles in os.walk(path_to_archive, followlinks=False):
```

Comment:
I find this part of the code difficult to understand. If I understand correctly it will manipulate `files_to_remove` and `dirs_to_remove_if_empty` by mutating the lists?

If this this would be much clearer the code to generate the files and dir paths lists was broken into a function, which was then used to instantiate the to lists.

Reply:
ok, yes I agree this part got a bit out of hand.. I'll break it up a bit